packages/engine/lib/error/src/frame.rs

```diff
@@ -333,15 +379,32 @@ where
 
 #[allow(clippy::used_underscore_binding)]
 impl FrameRepr {
-    fn unerase(&self) -> &dyn Context {
+    fn unerase(&self) -> &dyn Unerased {
         // SAFETY: Use vtable to attach C's native vtable for the right original type C.
         unsafe { (self.vtable.object_ref)(self) }
     }
 
-    fn downcast<C: Any>(&self) -> Option<NonNull<C>> {
-        let target = TypeId::of::<C>();
-        // SAFETY: Use vtable to attach C's native vtable for the right original type C.
-        let addr = unsafe { (self.vtable.object_downcast)(self, target) };
-        addr.map(NonNull::cast)
+    fn downcast<T: Any>(&self) -> Option<NonNull<T>> {
+        // TODO: Use tagged pointer to store the frame kind
+        //   see https://app.asana.com/0/0/1202366470755781/f
+        // SAFETY: Use vtable to attach C's native vtable for the right original type C. Casting
+        //   between `T` and `ContextRepr<T>`/`ErrorRepr<T>` is safe as those structs are
+        //   `repr(transparent)`.
+        unsafe {
+            if let Some(addr) = (self.vtable.object_downcast)(self, TypeId::of::<T>()) {
+                return Some(addr.cast());
+            }
+
+            if let Some(addr) = (self.vtable.object_downcast)(self, TypeId::of::<ContextRepr<T>>())
+            {
+                return Some(addr.cast());
+            }
+
+            #[cfg(feature = "std")]
+            if let Some(addr) = (self.vtable.object_downcast)(self, TypeId::of::<ErrorRepr<T>>()) {
```

Comment:
Depends on how you added the `Error` to the report. When adding it by `change_context` or `from_context`, it will ignore the provider implementation until `Error` has a blanket implementation for `Provider`. Then it will be used directly.